eth-providers/src/base-provider.ts

```diff
@@ -1233,6 +1233,21 @@ export abstract class BaseProvider extends AbstractProvider {
     }
   };
 
+  _isBlockFinalized = async (blockTag: BlockTag): boolean => {
+    const [finalizedHead, verifyingBlockHash] = await Promise.all([
+      this.api.rpc.chain.getFinalizedHead(),
+      this._getBlockHash(blockTag)
+    ]);
+
+    const [finalizedBlockNumber, verifyingBlockNumber] = (
+      await Promise.all([this.api.rpc.chain.getHeader(finalizedHead), this.api.rpc.chain.getHeader(verifyingBlockHash)])
```

Comment:
this is fine but in future we could cache some of those calls

Reply:
sounds good, we can make it work first and do some optimazation in the future

Reply:
I think the current performance is not too bad, already parallellized these fetches as much as possible. In the future we can do some more caching to make it O(1)